File: openadmin/spec/component.py
# SPDX-FileCopyrightText: 2026 OpenAdmin
#
# SPDX-License-Identifier: AGPL-3.0-or-later


from .action import ActionComponent
from .area_chart import AreaChartComponent
from .bar_chart import BarChartComponent
from .form import FormComponent
from .line_chart import LineChartComponent
from .markdown import MarkdownComponent
from .pie_chart import PieChartComponent
from .stat import StatComponent
from .table import TableComponent

type Component = (
    StatComponent
    | TableComponent
    | AreaChartComponent
    | BarChartComponent
    | LineChartComponent
    | PieChartComponent
    | ActionComponent
    | FormComponent
    | MarkdownComponent
)

File: client/src/schemas/component.ts
// SPDX-FileCopyrightText: 2026 OpenAdmin
//
// SPDX-License-Identifier: AGPL-3.0-or-later

import { z } from "zod"
import { actionComponentSchema } from "./action"
import { areaChartComponentSchema } from "./area-chart"
import { barChartComponentSchema } from "./bar-chart"
import { formComponentSchema } from "./form"
import { lineChartComponentSchema } from "./line-chart"
import { markdownComponentSchema } from "./markdown"
import { pieChartComponentSchema } from "./pie-chart"
import { statComponentSchema } from "./stat"
import { tableComponentSchema } from "./table"

export const componentSchema = z.discriminatedUnion("type", [
	statComponentSchema,
	tableComponentSchema,
	areaChartComponentSchema,
	barChartComponentSchema,
	lineChartComponentSchema,
	pieChartComponentSchema,
	actionComponentSchema,
	formComponentSchema,
	markdownComponentSchema,
])

export type Component = z.infer<typeof componentSchema>
